# 제품 이상여부 판별 프로젝트


### 데이터 읽어오기


In [1]:
import os
from pprint import pprint

import optuna
import numpy as np
import pandas as pd
#import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.tree import DecisionTreeClassifier
from lightgbm import LGBMClassifier
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier
from xgboost import XGBClassifier
from catboost import CatBoostClassifier
from sklearn.ensemble import AdaBoostClassifier

from sklearn.metrics import (
    accuracy_score,
    classification_report,
    confusion_matrix,
    f1_score,
    precision_score,
    recall_score,
)
from sklearn.model_selection import train_test_split
from tqdm import tqdm

/home/elicer/.local/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
import pandas as pd

RANDOM_STATE = 110

train_data = pd.read_csv("train_data_0827.csv")
test_data = pd.read_csv("test_data_0827.csv")

In [3]:
# dam, fill1, fill2 공통 변수
var_dam_fill = [
    'Receip_No_encoded',
    'Equipment_same_num',
    'PalletID_Collect_Result_encoded',
    'Production_Qty_Collect_Result',
    'WorkMode Collect Result'
]

In [4]:
# 전체 공통 변수
### train
var_all_train = [
    'target',
    'model_suffix_encoded',
    'cleaned_workorder_encoded'
]

### test
var_all_test = [
    'Set ID',
    'target',
    'model_suffix_encoded',
    'cleaned_workorder_encoded'
]

In [5]:
# '_Dam'을 포함하는 변수 선택
dam_variables = [var for var in train_data.columns if '_Dam' in var]

# train
final_columns_train = var_dam_fill + var_all_train + dam_variables
train_data_dam = train_data[final_columns_train]

# test 
final_columns_test = var_dam_fill + var_all_test + dam_variables
test_data_dam = test_data[final_columns_test]

In [6]:
# '_Fill1'을 포함하는 변수 선택
fill1_variables = [var for var in train_data.columns if '_Fill1' in var]

# train
final_columns_train = var_dam_fill + var_all_train + fill1_variables
train_data_fill1 = train_data[final_columns_train]

# test 
final_columns_test = var_dam_fill + var_all_test + fill1_variables
test_data_fill1 = test_data[final_columns_test]

In [7]:
# '_Fill2'을 포함하는 변수 선택
fill2_variables = [var for var in train_data.columns if '_Fill2' in var]

# train
final_columns_train = var_dam_fill + var_all_train + fill2_variables
train_data_fill2 = train_data[final_columns_train]

# test 
final_columns_test = var_dam_fill + var_all_test + fill2_variables
test_data_fill2 = test_data[final_columns_test]

In [8]:
# '_AutoClave'을 포함하는 변수 선택
autoclave_variables = [var for var in train_data.columns if '_AutoClave' in var]

# train
final_columns_train = var_all_train + autoclave_variables
train_data_autoclave = train_data[final_columns_train]

# test 
final_columns_test = var_all_test + autoclave_variables
test_data_autoclave = test_data[final_columns_test]

In [9]:
# 각 DataFrame의 칼럼 수 계산
num_columns_train_data = train_data.shape[1]
num_columns_train_data_dam = train_data_dam.shape[1]
num_columns_train_data_autoclave = train_data_autoclave.shape[1]
num_columns_train_data_fill1 = train_data_fill1.shape[1]
num_columns_train_data_fill2 = train_data_fill2.shape[1]

num_columns_test_data = test_data.shape[1]
num_columns_test_data_dam = test_data_dam.shape[1]
num_columns_test_data_autoclave = test_data_autoclave.shape[1]
num_columns_test_data_fill1 = test_data_fill1.shape[1]
num_columns_test_data_fill2 = test_data_fill2.shape[1]

# 각 DataFrame의 칼럼 수 출력
print("----train data-----")
print(f"train_data DataFrame의 칼럼 수: {num_columns_train_data}")
print(f"train_data_dam DataFrame의 칼럼 수: {num_columns_train_data_dam}")
print(f"train_data_autoclave DataFrame의 칼럼 수: {num_columns_train_data_autoclave}")
print(f"train_data_fill1 DataFrame의 칼럼 수: {num_columns_train_data_fill1}")
print(f"train_data_fill2 DataFrame의 칼럼 수: {num_columns_train_data_fill2}")
print("----test data-----")
print(f"test_data DataFrame의 칼럼 수: {num_columns_test_data}")
print(f"test_data_dam DataFrame의 칼럼 수: {num_columns_test_data_dam}")
print(f"test_data_autoclave DataFrame의 칼럼 수: {num_columns_test_data_autoclave}")
print(f"test_data_fill1 DataFrame의 칼럼 수: {num_columns_test_data_fill1}")
print(f"test_data_fill2 DataFrame의 칼럼 수: {num_columns_test_data_fill2}")

----train data-----
train_data DataFrame의 칼럼 수: 40
train_data_dam DataFrame의 칼럼 수: 23
train_data_autoclave DataFrame의 칼럼 수: 8
train_data_fill1 DataFrame의 칼럼 수: 14
train_data_fill2 DataFrame의 칼럼 수: 14
----test data-----
test_data DataFrame의 칼럼 수: 41
test_data_dam DataFrame의 칼럼 수: 24
test_data_autoclave DataFrame의 칼럼 수: 9
test_data_fill1 DataFrame의 칼럼 수: 15
test_data_fill2 DataFrame의 칼럼 수: 15


---

## Optuna

스레스홀드 0.3으로 맞춘상태에서 튜닝 진행한 것

In [10]:
RANDOM_STATE = 110
THRESHOLD = 0.3

In [11]:
import optuna
from catboost import CatBoostClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score

def objectiveCatBoost(trial, x_tr, y_tr, x_val, y_val):
    
    # 'Normal'과 'AbNormal'을 숫자로 변환
    y_tr = y_tr.map({'Normal': 0, 'AbNormal': 1})
    y_val = y_val.map({'Normal': 0, 'AbNormal': 1})

    param = {
        'iterations': trial.suggest_int('iterations', 800, 5000),
        'learning_rate': trial.suggest_float('learning_rate', 0.001, 0.2),
        'depth': trial.suggest_int('depth', 4, 13),
        'min_data_in_leaf': trial.suggest_int('min_data_in_leaf', 1, 5),
        'l2_leaf_reg': trial.suggest_float('l2_leaf_reg', 1, 10, log=True),
        'random_strength': trial.suggest_float('random_strength', 0, 10),
        'bagging_temperature': trial.suggest_float('bagging_temperature', 0.1, 10),
        'border_count': trial.suggest_int('border_count', 128, 300),
        'scale_pos_weight': trial.suggest_float('scale_pos_weight', 1, 10),
        'grow_policy': trial.suggest_categorical('grow_policy', ['SymmetricTree', 'Depthwise', 'Lossguide']),
        
        'random_seed': RANDOM_STATE,
        'eval_metric': 'F1',
        'logging_level': 'Silent',
        'boosting_type': 'Plain'
    }
    
    model = CatBoostClassifier(**param)
    model.fit(x_tr, y_tr)
    pred_proba = model.predict_proba(x_val)[:, 1]  # 양성 클래스 확률
    pred = (pred_proba >= THRESHOLD).astype(int)  # 스레드홀드에 따른 예측
    
    score = f1_score(y_val, pred, average="binary")
    
    return score

# 데이터셋 분할
x_train, x_val, y_train, y_val = train_test_split(
    train_data.drop("target", axis=1),  
    train_data["target"],               
    test_size=0.2,
    shuffle=True,
    random_state=RANDOM_STATE,
)

# 하이퍼 파라미터 튜닝
study = optuna.create_study(direction='maximize', sampler=optuna.samplers.TPESampler(seed=RANDOM_STATE))
study.optimize(lambda trial: objectiveCatBoost(trial, x_train, y_train, x_val, y_val), n_trials=500)

print('Best trial: score {}, \nparams {}'.format(study.best_trial.value, study.best_trial.params))

[I 2024-08-27 16:23:44,319] A new study created in memory with name: no-name-7e156289-746e-4856-b837-d1c9f3a0f94c
[I 2024-08-27 16:23:59,391] Trial 0 finished with value: 0.22463768115942032 and parameters: {'iterations': 1287, 'learning_rate': 0.13203262096748453, 'depth': 7, 'min_data_in_leaf': 4, 'l2_leaf_reg': 4.780689663026719, 'random_strength': 6.824631292225899, 'bagging_temperature': 9.554598960277284, 'border_count': 238, 'scale_pos_weight': 5.61649246971856, 'grow_policy': 'Depthwise'}. Best is trial 0 with value: 0.22463768115942032.
[I 2024-08-27 16:24:09,429] Trial 1 finished with value: 0.2197802197802198 and parameters: {'iterations': 1368, 'learning_rate': 0.0030110113610709835, 'depth': 8, 'min_data_in_leaf': 3, 'l2_leaf_reg': 1.0505036749294598, 'random_strength': 5.595100695350062, 'bagging_temperature': 0.8684052021612866, 'border_count': 198, 'scale_pos_weight': 3.8269447050797387, 'grow_policy': 'SymmetricTree'}. Best is trial 0 with value: 0.22463768115942032.
[

[I 2024-08-27 16:56:54,424] Trial 19 finished with value: 0.2121684867394696 and parameters: {'iterations': 4719, 'learning_rate': 0.09758867718691291, 'depth': 12, 'min_data_in_leaf': 1, 'l2_leaf_reg': 3.7370376286336344, 'random_strength': 2.175665029433702, 'bagging_temperature': 0.6012680873077576, 'border_count': 284, 'scale_pos_weight': 1.0674410565053178, 'grow_policy': 'SymmetricTree'}. Best is trial 15 with value: 0.23127463863337713.
[I 2024-08-27 17:00:18,616] Trial 20 finished with value: 0.2222222222222222 and parameters: {'iterations': 3562, 'learning_rate': 0.053526595531264765, 'depth': 12, 'min_data_in_leaf': 2, 'l2_leaf_reg': 7.481413232538302, 'random_strength': 0.9911188121443996, 'bagging_temperature': 0.10828299341959258, 'border_count': 251, 'scale_pos_weight': 6.59149106322385, 'grow_policy': 'SymmetricTree'}. Best is trial 15 with value: 0.23127463863337713.
[I 2024-08-27 17:02:03,502] Trial 21 finished with value: 0.22733812949640286 and parameters: {'iteratio

[I 2024-08-27 17:32:51,143] Trial 38 finished with value: 0.23303834808259588 and parameters: {'iterations': 2608, 'learning_rate': 0.015852652643708193, 'depth': 13, 'min_data_in_leaf': 5, 'l2_leaf_reg': 1.1977538589999803, 'random_strength': 0.5542711178402808, 'bagging_temperature': 3.3435156358003844, 'border_count': 244, 'scale_pos_weight': 2.570526472348657, 'grow_policy': 'Depthwise'}. Best is trial 23 with value: 0.23809523809523808.
[I 2024-08-27 17:34:00,861] Trial 39 finished with value: 0.23738062755798087 and parameters: {'iterations': 2519, 'learning_rate': 0.013423575069977704, 'depth': 11, 'min_data_in_leaf': 5, 'l2_leaf_reg': 2.0700523506585675, 'random_strength': 4.0687045842651575, 'bagging_temperature': 3.5120763997019333, 'border_count': 244, 'scale_pos_weight': 2.764111873681697, 'grow_policy': 'Depthwise'}. Best is trial 23 with value: 0.23809523809523808.
[I 2024-08-27 17:35:01,126] Trial 40 finished with value: 0.22571428571428573 and parameters: {'iterations':

[I 2024-08-27 17:48:36,208] Trial 57 finished with value: 0.2332361516034985 and parameters: {'iterations': 1982, 'learning_rate': 0.17718295892962332, 'depth': 10, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.5801815325645934, 'random_strength': 6.136749247260928, 'bagging_temperature': 9.883368461420794, 'border_count': 182, 'scale_pos_weight': 2.8139932654435995, 'grow_policy': 'Depthwise'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 17:49:11,066] Trial 58 finished with value: 0.2303206997084548 and parameters: {'iterations': 1976, 'learning_rate': 0.15881570419231786, 'depth': 10, 'min_data_in_leaf': 3, 'l2_leaf_reg': 1.3784483348035996, 'random_strength': 6.29245188626566, 'bagging_temperature': 9.998509668872495, 'border_count': 186, 'scale_pos_weight': 1.9548867482447703, 'grow_policy': 'Depthwise'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 17:49:29,979] Trial 59 finished with value: 0.22720897615708274 and parameters: {'iterations': 1714, 

[I 2024-08-27 18:02:10,390] Trial 76 finished with value: 0.2323529411764706 and parameters: {'iterations': 3009, 'learning_rate': 0.18420394192054626, 'depth': 11, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.0607854971936905, 'random_strength': 9.359198366145588, 'bagging_temperature': 6.724566474301709, 'border_count': 134, 'scale_pos_weight': 2.1102551123877404, 'grow_policy': 'Depthwise'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 18:02:43,256] Trial 77 finished with value: 0.2193211488250653 and parameters: {'iterations': 2847, 'learning_rate': 0.022161727195238734, 'depth': 12, 'min_data_in_leaf': 3, 'l2_leaf_reg': 1.135557928361943, 'random_strength': 7.685094026573509, 'bagging_temperature': 7.186938983962753, 'border_count': 158, 'scale_pos_weight': 2.3886307892630025, 'grow_policy': 'Lossguide'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 18:03:17,510] Trial 78 finished with value: 0.23087621696801108 and parameters: {'iterations': 2216,

[I 2024-08-27 18:14:32,469] Trial 95 finished with value: 0.22816901408450704 and parameters: {'iterations': 1628, 'learning_rate': 0.17256661610856527, 'depth': 11, 'min_data_in_leaf': 5, 'l2_leaf_reg': 2.9543545299353697, 'random_strength': 3.3931189941110205, 'bagging_temperature': 4.345653674471528, 'border_count': 250, 'scale_pos_weight': 2.255327458327764, 'grow_policy': 'Lossguide'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 18:14:45,324] Trial 96 finished with value: 0.23076923076923078 and parameters: {'iterations': 1131, 'learning_rate': 0.16614881074985408, 'depth': 12, 'min_data_in_leaf': 3, 'l2_leaf_reg': 1.1212755602489002, 'random_strength': 4.493984005638348, 'bagging_temperature': 3.5382897967013487, 'border_count': 260, 'scale_pos_weight': 2.549434274305289, 'grow_policy': 'Lossguide'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 18:15:17,811] Trial 97 finished with value: 0.23936170212765956 and parameters: {'iterations': 264

[I 2024-08-27 18:28:45,778] Trial 114 finished with value: 0.1818705035971223 and parameters: {'iterations': 1804, 'learning_rate': 0.0010696753390855972, 'depth': 11, 'min_data_in_leaf': 5, 'l2_leaf_reg': 1.0813801200414104, 'random_strength': 4.279062761037138, 'bagging_temperature': 4.781787033258922, 'border_count': 138, 'scale_pos_weight': 6.208484481869225, 'grow_policy': 'Depthwise'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 18:29:10,428] Trial 115 finished with value: 0.22733812949640286 and parameters: {'iterations': 2041, 'learning_rate': 0.18073352378077137, 'depth': 8, 'min_data_in_leaf': 5, 'l2_leaf_reg': 1.2659474674266586, 'random_strength': 1.9325566904916842, 'bagging_temperature': 8.135946019305711, 'border_count': 147, 'scale_pos_weight': 2.1261871578819806, 'grow_policy': 'Depthwise'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 18:29:58,498] Trial 116 finished with value: 0.21145374449339208 and parameters: {'iterations': 

[I 2024-08-27 18:43:14,751] Trial 133 finished with value: 0.22416302765647744 and parameters: {'iterations': 1712, 'learning_rate': 0.11967999514039458, 'depth': 10, 'min_data_in_leaf': 4, 'l2_leaf_reg': 2.1864944213258966, 'random_strength': 3.5597534786792195, 'bagging_temperature': 9.349274333215858, 'border_count': 233, 'scale_pos_weight': 2.5775751585587647, 'grow_policy': 'Depthwise'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 18:44:33,957] Trial 134 finished with value: 0.23391812865497078 and parameters: {'iterations': 4289, 'learning_rate': 0.07587485111055509, 'depth': 11, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.2065801942744343, 'random_strength': 2.6020892690803907, 'bagging_temperature': 4.380111498695108, 'border_count': 237, 'scale_pos_weight': 2.801600176976141, 'grow_policy': 'Depthwise'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 18:45:46,524] Trial 135 finished with value: 0.22418879056047197 and parameters: {'iterations':

[I 2024-08-27 19:00:24,593] Trial 152 finished with value: 0.21067821067821071 and parameters: {'iterations': 2936, 'learning_rate': 0.0812537841333874, 'depth': 10, 'min_data_in_leaf': 4, 'l2_leaf_reg': 2.399865485209695, 'random_strength': 0.48777407893905916, 'bagging_temperature': 3.0558479655389443, 'border_count': 262, 'scale_pos_weight': 1.9235330037221972, 'grow_policy': 'Lossguide'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 19:01:01,308] Trial 153 finished with value: 0.22451994091580504 and parameters: {'iterations': 3392, 'learning_rate': 0.05973866858245467, 'depth': 10, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.1344948092760334, 'random_strength': 1.0464655036248385, 'bagging_temperature': 3.6624551197409705, 'border_count': 271, 'scale_pos_weight': 1.5290844281572702, 'grow_policy': 'Lossguide'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 19:01:46,084] Trial 154 finished with value: 0.22031473533619458 and parameters: {'iterations

[I 2024-08-27 19:15:49,098] Trial 171 finished with value: 0.22288261515601782 and parameters: {'iterations': 1769, 'learning_rate': 0.18517537496362485, 'depth': 10, 'min_data_in_leaf': 1, 'l2_leaf_reg': 1.0093544798458498, 'random_strength': 7.229642083528982, 'bagging_temperature': 6.082955142954463, 'border_count': 262, 'scale_pos_weight': 2.5476480852640107, 'grow_policy': 'Depthwise'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 19:16:23,505] Trial 172 finished with value: 0.2143906020558003 and parameters: {'iterations': 1933, 'learning_rate': 0.17966509760581134, 'depth': 10, 'min_data_in_leaf': 2, 'l2_leaf_reg': 1.0591877020526757, 'random_strength': 6.863377382736995, 'bagging_temperature': 5.279726233540197, 'border_count': 267, 'scale_pos_weight': 2.8315028160183973, 'grow_policy': 'Depthwise'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 19:16:54,076] Trial 173 finished with value: 0.2346041055718475 and parameters: {'iterations': 15

[I 2024-08-27 19:28:20,801] Trial 190 finished with value: 0.23346303501945528 and parameters: {'iterations': 1639, 'learning_rate': 0.013161822487085466, 'depth': 11, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.092684992639209, 'random_strength': 0.8438080903659597, 'bagging_temperature': 9.699068120336545, 'border_count': 266, 'scale_pos_weight': 3.4309954554799003, 'grow_policy': 'Depthwise'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 19:29:04,119] Trial 191 finished with value: 0.23468057366362452 and parameters: {'iterations': 1776, 'learning_rate': 0.010688555115119888, 'depth': 10, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.1554421848371137, 'random_strength': 0.6302191714735055, 'bagging_temperature': 9.390239268090683, 'border_count': 263, 'scale_pos_weight': 2.584164325353061, 'grow_policy': 'Depthwise'}. Best is trial 53 with value: 0.24313725490196078.
[I 2024-08-27 19:29:52,893] Trial 192 finished with value: 0.23379923761118165 and parameters: {'iterations'

[I 2024-08-27 19:39:32,560] Trial 209 finished with value: 0.24561403508771928 and parameters: {'iterations': 1627, 'learning_rate': 0.009687449220309025, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.046304181584105, 'random_strength': 0.4298254458723779, 'bagging_temperature': 8.104944759743836, 'border_count': 202, 'scale_pos_weight': 2.7255430696738854, 'grow_policy': 'Depthwise'}. Best is trial 205 with value: 0.2503160556257901.
[I 2024-08-27 19:40:02,349] Trial 210 finished with value: 0.22448979591836737 and parameters: {'iterations': 1636, 'learning_rate': 0.0016286528434968275, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.0435818797424667, 'random_strength': 0.4649272432866185, 'bagging_temperature': 8.119552977957515, 'border_count': 206, 'scale_pos_weight': 2.6589814475716484, 'grow_policy': 'Depthwise'}. Best is trial 205 with value: 0.2503160556257901.
[I 2024-08-27 19:40:30,740] Trial 211 finished with value: 0.24849215922798557 and parameters: {'iterations'

[I 2024-08-27 19:48:31,872] Trial 228 finished with value: 0.23407022106631986 and parameters: {'iterations': 1405, 'learning_rate': 0.015388059618514015, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.0024004821223722, 'random_strength': 0.35104656648440097, 'bagging_temperature': 7.76498282820862, 'border_count': 198, 'scale_pos_weight': 2.8272217186812525, 'grow_policy': 'Depthwise'}. Best is trial 205 with value: 0.2503160556257901.
[I 2024-08-27 19:48:54,922] Trial 229 finished with value: 0.22670025188916876 and parameters: {'iterations': 1589, 'learning_rate': 0.0054803998723713995, 'depth': 8, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.059693354439661, 'random_strength': 0.3003956210583507, 'bagging_temperature': 8.178042178382512, 'border_count': 195, 'scale_pos_weight': 2.5938469189871887, 'grow_policy': 'Depthwise'}. Best is trial 205 with value: 0.2503160556257901.
[I 2024-08-27 19:49:27,115] Trial 230 finished with value: 0.24521072796934867 and parameters: {'iterations'

[I 2024-08-27 19:57:44,887] Trial 247 finished with value: 0.2219679633867277 and parameters: {'iterations': 1613, 'learning_rate': 0.007843726003135928, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.0000477061869864, 'random_strength': 0.7776343343273541, 'bagging_temperature': 7.8077121725108745, 'border_count': 215, 'scale_pos_weight': 2.860473986879183, 'grow_policy': 'SymmetricTree'}. Best is trial 205 with value: 0.2503160556257901.
[I 2024-08-27 19:58:18,459] Trial 248 finished with value: 0.220708446866485 and parameters: {'iterations': 1760, 'learning_rate': 0.015512467979218736, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.1413455286410195, 'random_strength': 0.23541123051491888, 'bagging_temperature': 8.665452949630538, 'border_count': 188, 'scale_pos_weight': 2.6198134599406666, 'grow_policy': 'Depthwise'}. Best is trial 205 with value: 0.2503160556257901.
[I 2024-08-27 19:58:46,559] Trial 249 finished with value: 0.25405405405405407 and parameters: {'iteration

[I 2024-08-27 20:06:44,032] Trial 266 finished with value: 0.23058823529411762 and parameters: {'iterations': 1669, 'learning_rate': 0.00485040745260877, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.069973727411686, 'random_strength': 0.37155085046946884, 'bagging_temperature': 8.532916640919431, 'border_count': 216, 'scale_pos_weight': 2.876470299339839, 'grow_policy': 'Depthwise'}. Best is trial 249 with value: 0.25405405405405407.
[I 2024-08-27 20:07:10,599] Trial 267 finished with value: 0.23906083244397008 and parameters: {'iterations': 1461, 'learning_rate': 0.00889170860384828, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.035387698948925, 'random_strength': 0.7516752288426394, 'bagging_temperature': 7.5687111272173295, 'border_count': 196, 'scale_pos_weight': 3.2523013538211822, 'grow_policy': 'Depthwise'}. Best is trial 249 with value: 0.25405405405405407.
[I 2024-08-27 20:07:41,219] Trial 268 finished with value: 0.2360248447204969 and parameters: {'iterations': 

[I 2024-08-27 20:15:17,240] Trial 285 finished with value: 0.23138297872340427 and parameters: {'iterations': 1594, 'learning_rate': 0.013445865505665624, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.1746507706058218, 'random_strength': 0.5102144047190796, 'bagging_temperature': 8.472118360109212, 'border_count': 192, 'scale_pos_weight': 2.5765651081668093, 'grow_policy': 'Depthwise'}. Best is trial 249 with value: 0.25405405405405407.
[I 2024-08-27 20:15:43,076] Trial 286 finished with value: 0.2174721189591078 and parameters: {'iterations': 1669, 'learning_rate': 0.005164735382075566, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.0533472491054965, 'random_strength': 6.408888572589957, 'bagging_temperature': 8.285759081826644, 'border_count': 202, 'scale_pos_weight': 3.0035547450123445, 'grow_policy': 'Depthwise'}. Best is trial 249 with value: 0.25405405405405407.
[I 2024-08-27 20:16:10,972] Trial 287 finished with value: 0.22865013774104684 and parameters: {'iterations'

[I 2024-08-27 20:22:09,655] Trial 304 finished with value: 0.23876765083440307 and parameters: {'iterations': 884, 'learning_rate': 0.026826604628612763, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.151102193118165, 'random_strength': 0.5000251992798208, 'bagging_temperature': 8.85672771473291, 'border_count': 198, 'scale_pos_weight': 2.9856687613045776, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 20:22:27,542] Trial 305 finished with value: 0.24085005903187723 and parameters: {'iterations': 960, 'learning_rate': 0.01612104137179998, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.1131219261603769, 'random_strength': 0.7640130018557733, 'bagging_temperature': 8.985892930996748, 'border_count': 186, 'scale_pos_weight': 2.834255912385405, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 20:22:43,997] Trial 306 finished with value: 0.193029490616622 and parameters: {'iterations': 827, '

[I 2024-08-27 20:29:19,400] Trial 323 finished with value: 0.23591087811271297 and parameters: {'iterations': 1634, 'learning_rate': 0.007671210021902396, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.1061661874899775, 'random_strength': 0.3801178162713374, 'bagging_temperature': 8.163091657706461, 'border_count': 206, 'scale_pos_weight': 2.3335076643538852, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 20:29:46,793] Trial 324 finished with value: 0.23451910408432147 and parameters: {'iterations': 1383, 'learning_rate': 0.01528567792335197, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.043295690027727, 'random_strength': 0.22075902521170546, 'bagging_temperature': 8.004737682939343, 'border_count': 200, 'scale_pos_weight': 2.7383696502762254, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 20:30:14,233] Trial 325 finished with value: 0.24585635359116023 and parameters: {'iterations':

[I 2024-08-27 20:37:11,290] Trial 342 finished with value: 0.25392296718972895 and parameters: {'iterations': 1658, 'learning_rate': 0.009828228606361403, 'depth': 8, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.7522372287376393, 'random_strength': 0.7004835422689379, 'bagging_temperature': 8.527676522275454, 'border_count': 182, 'scale_pos_weight': 1.8354720824020405, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 20:37:33,693] Trial 343 finished with value: 0.22222222222222224 and parameters: {'iterations': 1628, 'learning_rate': 0.009301932945519334, 'depth': 8, 'min_data_in_leaf': 4, 'l2_leaf_reg': 4.667402508085391, 'random_strength': 0.6529326120287238, 'bagging_temperature': 8.436487320397768, 'border_count': 189, 'scale_pos_weight': 1.5770678685811579, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 20:37:54,528] Trial 344 finished with value: 0.20386643233743407 and parameters: {'iterations':

[I 2024-08-27 20:44:32,641] Trial 361 finished with value: 0.24701195219123503 and parameters: {'iterations': 1480, 'learning_rate': 0.01076921301837019, 'depth': 8, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.1365597466382134, 'random_strength': 0.993613910144667, 'bagging_temperature': 9.016341501209162, 'border_count': 193, 'scale_pos_weight': 2.18334155274338, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 20:44:57,386] Trial 362 finished with value: 0.23638778220451528 and parameters: {'iterations': 1641, 'learning_rate': 0.01413575078561668, 'depth': 8, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.1450702295537312, 'random_strength': 0.0018889827460157327, 'bagging_temperature': 8.954668639090844, 'border_count': 194, 'scale_pos_weight': 2.3112709593925085, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 20:45:20,681] Trial 363 finished with value: 0.2259983007646559 and parameters: {'iterations': 1

[I 2024-08-27 20:52:12,914] Trial 380 finished with value: 0.21517553793884486 and parameters: {'iterations': 1408, 'learning_rate': 0.001221645990326703, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.1308083551984378, 'random_strength': 0.7252673817465861, 'bagging_temperature': 7.8356012202890595, 'border_count': 201, 'scale_pos_weight': 2.6356240576193644, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 20:52:25,629] Trial 381 finished with value: 0.2203567681007345 and parameters: {'iterations': 1161, 'learning_rate': 0.020085885874839372, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.0367102594307975, 'random_strength': 0.20515914589238193, 'bagging_temperature': 8.736159079770236, 'border_count': 205, 'scale_pos_weight': 3.5206853483330898, 'grow_policy': 'SymmetricTree'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 20:52:53,393] Trial 382 finished with value: 0.25100133511348466 and parameters: {'iterat

[I 2024-08-27 20:59:48,444] Trial 399 finished with value: 0.2396804260985353 and parameters: {'iterations': 1386, 'learning_rate': 0.011529226043918941, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.4578806998077969, 'random_strength': 0.74775730393096, 'bagging_temperature': 9.107790711830825, 'border_count': 194, 'scale_pos_weight': 2.253480819307326, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 21:00:16,653] Trial 400 finished with value: 0.23380281690140842 and parameters: {'iterations': 1494, 'learning_rate': 0.015638133547630873, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.4728637595864875, 'random_strength': 0.5012613808311333, 'bagging_temperature': 9.56265339335505, 'border_count': 191, 'scale_pos_weight': 2.0298430233066584, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 21:00:39,106] Trial 401 finished with value: 0.24043715846994534 and parameters: {'iterations': 123

[I 2024-08-27 21:07:46,132] Trial 418 finished with value: 0.24431818181818182 and parameters: {'iterations': 1523, 'learning_rate': 0.008716879990254882, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.6317199520048327, 'random_strength': 0.9780829086346614, 'bagging_temperature': 8.334048694486551, 'border_count': 213, 'scale_pos_weight': 1.9507875025778998, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 21:08:57,043] Trial 419 finished with value: 0.24501424501424504 and parameters: {'iterations': 3918, 'learning_rate': 0.004394933037650556, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.9304236234211147, 'random_strength': 1.0195537432514468, 'bagging_temperature': 8.303444934084194, 'border_count': 215, 'scale_pos_weight': 1.8076160963353731, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 21:09:24,487] Trial 420 finished with value: 0.20996441281138795 and parameters: {'iterations'

[I 2024-08-27 21:22:26,575] Trial 437 finished with value: 0.24802110817941955 and parameters: {'iterations': 1583, 'learning_rate': 0.011093546868732148, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 2.046244245532402, 'random_strength': 1.1915996508421867, 'bagging_temperature': 7.927622803227754, 'border_count': 200, 'scale_pos_weight': 2.1676394509499777, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 21:22:52,754] Trial 438 finished with value: 0.2365930599369085 and parameters: {'iterations': 1573, 'learning_rate': 0.0075139188638014935, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 2.098374794323526, 'random_strength': 1.6263367333993508, 'bagging_temperature': 7.77583826694708, 'border_count': 210, 'scale_pos_weight': 1.7116741461521585, 'grow_policy': 'Depthwise'}. Best is trial 294 with value: 0.2540880503144654.
[I 2024-08-27 21:23:18,858] Trial 439 finished with value: 0.2289348171701113 and parameters: {'iterations': 15

[I 2024-08-27 21:31:30,674] Trial 456 finished with value: 0.2056338028169014 and parameters: {'iterations': 1343, 'learning_rate': 0.0012562557901038638, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.994237494778404, 'random_strength': 1.6025815782223216, 'bagging_temperature': 8.563187857230893, 'border_count': 205, 'scale_pos_weight': 2.475438716990256, 'grow_policy': 'Depthwise'}. Best is trial 445 with value: 0.25439783491204326.
[I 2024-08-27 21:31:56,633] Trial 457 finished with value: 0.2301369863013699 and parameters: {'iterations': 1474, 'learning_rate': 0.00701463182084154, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 2.1402563645240833, 'random_strength': 0.716197293369808, 'bagging_temperature': 8.263012837839726, 'border_count': 208, 'scale_pos_weight': 2.0964471051316522, 'grow_policy': 'Depthwise'}. Best is trial 445 with value: 0.25439783491204326.
[I 2024-08-27 21:33:03,191] Trial 458 finished with value: 0.23927178153446033 and parameters: {'iterations': 3

[I 2024-08-27 21:41:47,159] Trial 475 finished with value: 0.22879177377892032 and parameters: {'iterations': 1678, 'learning_rate': 0.01098332178620216, 'depth': 8, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.1660386607806963, 'random_strength': 0.8010921289728666, 'bagging_temperature': 8.308568890009365, 'border_count': 203, 'scale_pos_weight': 2.263383040111784, 'grow_policy': 'Depthwise'}. Best is trial 445 with value: 0.25439783491204326.
[I 2024-08-27 21:42:13,314] Trial 476 finished with value: 0.23249299719887956 and parameters: {'iterations': 1328, 'learning_rate': 0.017164266260144008, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.2921101392056547, 'random_strength': 0.42725170763025827, 'bagging_temperature': 7.469263354526021, 'border_count': 212, 'scale_pos_weight': 1.9786097150119706, 'grow_policy': 'Depthwise'}. Best is trial 445 with value: 0.25439783491204326.
[I 2024-08-27 21:42:33,403] Trial 477 finished with value: 0.24203821656050958 and parameters: {'iterations

[I 2024-08-27 21:50:17,126] Trial 494 finished with value: 0.2305389221556886 and parameters: {'iterations': 1587, 'learning_rate': 0.13439983724476517, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.1969584187016609, 'random_strength': 1.467383087383223, 'bagging_temperature': 7.632585287058977, 'border_count': 212, 'scale_pos_weight': 1.920515310538692, 'grow_policy': 'Depthwise'}. Best is trial 445 with value: 0.25439783491204326.
[I 2024-08-27 21:50:48,379] Trial 495 finished with value: 0.23458646616541354 and parameters: {'iterations': 1832, 'learning_rate': 0.00324049483923918, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 1.2899210593368493, 'random_strength': 1.194152012828178, 'bagging_temperature': 7.92764605280196, 'border_count': 219, 'scale_pos_weight': 2.07426719138714, 'grow_policy': 'Depthwise'}. Best is trial 445 with value: 0.25439783491204326.
[I 2024-08-27 21:51:16,146] Trial 496 finished with value: 0.2 and parameters: {'iterations': 1737, 'learning_rate'

Best trial: score 0.25439783491204326, 
params {'iterations': 1349, 'learning_rate': 0.012526639112437014, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 2.245006704049574, 'random_strength': 0.6922797458293842, 'bagging_temperature': 8.230635636022027, 'border_count': 211, 'scale_pos_weight': 2.0709015241138236, 'grow_policy': 'Depthwise'}


1. 0.25439783491204326

params {'iterations': 1349, 'learning_rate': 0.012526639112437014, 'depth': 9, 'min_data_in_leaf': 4, 'l2_leaf_reg': 2.245006704049574, 'random_strength': 0.6922797458293842, 'bagging_temperature': 8.230635636022027, 'border_count': 211, 'scale_pos_weight': 2.0709015241138236, 'grow_policy': 'Depthwise'}

In [11]:
# 모델 설정 및 하이퍼파라미터
models = {
    'et': ExtraTreesClassifier(),
    'rf': RandomForestClassifier(),
    'cat': CatBoostClassifier(),
    'lgbm': LGBMClassifier(),
    'xgb': XGBClassifier(),
    'dt': DecisionTreeClassifier()
}

def train_and_evaluate_model(model_name, data, **params):
    if model_name not in models:
        print(f"{model_name}은(는) 지원되지 않는 모델입니다.")
        return
    
    # 데이터셋 분할
    x_train, x_val, y_train, y_val = train_test_split(
        data.drop("target", axis=1),
        data["target"].map({'Normal': 0, 'AbNormal': 1}),
        test_size=0.2,
        shuffle=True,
        random_state=RANDOM_STATE,
    )

    # 모델 선택
    model = models[model_name]

    # 하이퍼파라미터 설정
    model.set_params(**params)

    # 모델 학습
    model.fit(x_train, y_train)

    # 데이터 이름을 자동으로 추출하기 위한 래퍼 함수
    data_name = [name for name in globals() if globals()[name] is data][0]

    # 예측
    y_val_pred_proba = model.predict_proba(x_val)[:, 1]  # 양성 클래스 확률
    y_val_pred = (y_val_pred_proba >= THRESHOLD).astype(int)  # 스레드홀드에 따른 예측

    # 평가지표 계산
    f1 = f1_score(y_val, y_val_pred, average="binary")
    accuracy = accuracy_score(y_val, y_val_pred)
    precision = precision_score(y_val, y_val_pred)
    recall = recall_score(y_val, y_val_pred)
    conf_matrix = confusion_matrix(y_val, y_val_pred)
    
    # 평가지표 계산
    f1 = f1_score(y_val, y_val_pred, average="binary")
    accuracy = accuracy_score(y_val, y_val_pred)
    precision = precision_score(y_val, y_val_pred)
    recall = recall_score(y_val, y_val_pred)
    conf_matrix = confusion_matrix(y_val, y_val_pred)
    
    # 결과 출력
    print(f'{model_name} 모델이 {data_name} 데이터로 학습한 결과:')
    print(f'F1 Score: {f1}')
    print('---')
    print('Confusion Matrix:')
    print(conf_matrix)
    print('---')
    print(f'Accuracy: {accuracy}')
    print(f'Precision: {precision}')
    print(f'Recall: {recall}')
    print('\n')

In [12]:
train_and_evaluate_model(
    'cat', train_data,
    iterations=1349,
    learning_rate=0.012526639112437014,
    depth=9,
    min_data_in_leaf=4,
    l2_leaf_reg=2.245006704049574,
    random_strength=0.6922797458293842,
    bagging_temperature=8.230635636022027,
    border_count=211,
    scale_pos_weight=2.0709015241138236,
    grow_policy='Depthwise',
    random_seed=RANDOM_STATE,
    eval_metric='F1',
    logging_level='Silent',
    boosting_type='Plain'
)

cat 모델이 train_data 데이터로 학습한 결과:
F1 Score: 0.23880597014925375
---
Confusion Matrix:
[[7453  209]
 [ 352   88]]
---
Accuracy: 0.9307578375709701
Precision: 0.2962962962962963
Recall: 0.2


